### PR TITLE
Bugfix/#47

### PR DESCRIPTION
**Pull request by leborgneromain**
Fixed the bug on the sub-labels + refacto of the refreshHighlightedNodes function

### CODE REVIEW

File: web/ui/src/app/lineage/graph/graph-builder.ts [hunk 1]
```diff
@@ -34,13 +37,74 @@ export function lineageToGraph(lineage: IDataLineage,
...
+/**
+ * Get label from an operation
+ * 
+ * @param operation the operation where the label should be retrieved
+ * @param expressionRenderService service to render expressions
+ * 
+ * @returns the label from the operation
+ */
+export function getLabel(operation : (IOperation), expressionRenderService : ExpressionRenderService){
+    let label = operation.mainProps.name + "\n";
```
Comment (wajda) [inline web/ui/src/app/lineage/graph/graph-builder.ts]: Semicolons. Unless you have a strong opinion why we _should_ use semicolons I'd say we rather don't.
While in JS world, there is a never ending battle about if semicolons are good or evil, in TS however they make even less sense. Even in JS the cases that make ASI to behave unexpectedly are rare and strange enough to not belong to a good coding style anyway. But in TS the compiler will catch a good chuck of those leaving even less room for ASI related mistakes. There is a good article about it - https://medium.com/@eugenkiss/dont-use-semicolons-in-typescript-474ccfe4bdb3

File: web/ui/src/app/lineage/graph/graph-builder.ts [hunk 2]
```diff
@@ -34,13 +37,74 @@ export function lineageToGraph(lineage: IDataLineage,
...
+ * @param expressionRenderService service to render expressions
+ * 
+ * @returns the label from the operation
+ */
+export function getLabel(operation : (IOperation), expressionRenderService : ExpressionRenderService){
+    let label = operation.mainProps.name + "\n"; 
+    switch(typeOfOperation(operation)){
+        case "Aggregate":
+            let aggregations = (<any> operation).aggregations;
+            label += Object.keys(aggregations).join(",");break;
```
Comment (wajda) [inline web/ui/src/app/lineage/graph/graph-builder.ts]: `break` should be on a new line.
Having it in the end of the previous line makes it very easy to overlook when reading the code. If you want to keep the `switch `-block  more compact consider moving it in a separate function and use `case ... : return ...` pattern, concatenating strings outside.

File: web/ui/src/app/lineage/graph/graph-builder.ts [hunk 3]
```diff
@@ -34,13 +37,74 @@ export function lineageToGraph(lineage: IDataLineage,
...
+        case "Join":
+            let joins = (<IJoin> operation).condition;
```
Comment (wajda) [inline web/ui/src/app/lineage/graph/graph-builder.ts]: Try to find more meaningful names for variables. The names like `joins` or `filters` don't make it obvious that they contain a value of a singular expression type.

File: web/ui/src/app/lineage/graph/graph-builder.ts [hunk 4]
```diff
@@ -34,13 +37,74 @@ export function lineageToGraph(lineage: IDataLineage,
...
+            label += expressionRenderService.getText(joins);break;
+        case "Filter":
+            let filters = (<IFilter> operation).condition; 
+            label += expressionRenderService.getText(filters);break;
+        case "Projection":
+            // TODO : Define what to display for the second label
+            break;
+        case "Sort":
+        let sorts = (<ISort> operation).orders;
```
Comment (wajda) [inline web/ui/src/app/lineage/graph/graph-builder.ts]: indentation

File: web/ui/src/app/lineage/graph/graph-builder.ts (hunk 1 above)
Comment (vackosar) [reply web/ui/src/app/lineage/graph/graph-builder.ts]: I prefer semicollons in TS, but I  dont mind either way.

Comment (leborgneromain) [reply web/ui/src/app/lineage/graph/graph-builder.ts]: As you guys prefer, I have no preference.

File: web/ui/src/app/lineage/graph/graph-builder.ts (hunk 2 above)
Comment (leborgneromain) [reply web/ui/src/app/lineage/graph/graph-builder.ts]: ok ok. I am used to an autoformat on save that was obviously not here.

File: web/ui/src/app/lineage/graph/graph-builder.ts (hunk 3 above)
Comment (leborgneromain) [reply web/ui/src/app/lineage/graph/graph-builder.ts]: Renammed

File: web/ui/src/app/lineage/graph/graph-builder.ts [hunk 5]
```diff
@@ -34,13 +37,74 @@ export function lineageToGraph(lineage: IDataLineage,
...
+            label += expressionRenderService.getText(filters);break;
+        case "Projection":
+            // TODO : Define what to display for the second label
+            break;
+        case "Sort":
+        let sorts = (<ISort> operation).orders; 
+        label += sorts[0].direction + " " + expressionRenderService.getText(sorts[0]);break;
+        case "Read":
+            let read = (<IRead> operation).sources;
+            label += getFileName(read[0].path);break;
```
[the remark this answers is not in the document]
Comment (leborgneromain) [reply web/ui/src/app/lineage/graph/graph-builder.ts]: Added the path for the write operation.

File: web/ui/src/app/lineage/graph/graph-builder.ts [hunk 6]
```diff
@@ -34,13 +37,74 @@ export function lineageToGraph(lineage: IDataLineage,
...
 }
+/**
+ * Get label from an operation
+ * 
+ * @param operation the operation where the label should be retrieved
+ * @param expressionRenderService service to render expressions
+ * 
+ * @returns the label from the operation
+ */
+export function getLabel(operation : (IOperation), expressionRenderService : ExpressionRenderService){
```
[the remark this answers is not in the document]
Comment (leborgneromain) [reply web/ui/src/app/lineage/graph/graph-builder.ts]: yes they are. Removed

File: web/ui/src/app/lineage/graph/graph-builder.ts [hunk 7]
```diff
@@ -34,13 +37,74 @@ export function lineageToGraph(lineage: IDataLineage,
...
         hiddenOpIds: string[] = _.map(hiddenOperations, "mainProps.id"),
-        visibleNodes = visibleOperations.map(op => new RegularVisNode(op)),
+        visibleNodes = visibleOperations.map(op => new RegularVisNode(op, getLabel(op, expressionRenderService))),
         visibleEdges = createVisibleEdges(lineage, hiddenOpIds)
 
     return new VisModel(
         new vis.DataSet<VisNode>(visibleNodes),
         new vis.DataSet<VisEdge>(visibleEdges))
 }
+/**
```
[the remark this answers is not in the document]
Comment (leborgneromain) [reply web/ui/src/app/lineage/graph/graph-builder.ts]: Well, I am trying to comment every of my function even tho I do agree that these one are pretty straight forward functions.

**Comment by leborgneromain** [comment]
Corrected code after your code review.

File: web/ui/src/app/lineage/graph/graph-builder.ts [hunk 8]
```diff
@@ -34,13 +37,82 @@ export function lineageToGraph(lineage: IDataLineage,
...
+}
+
+/**
+ * Takes the first maxSize characters and add ellipsis to it
+ * @param label the label to format
+ * @param maxSize the size of the final string
+ * 
+ * return the input label formated
+ */
+function formatLabel(label: string, maxSize: number = 50): string{
```
Comment (wajda) [inline web/ui/src/app/lineage/graph/graph-builder.ts]: This function is redundant. There is `truncate` method in LoDash that does exactly that. And it's also a little bit more accurate at placing "..." when the original string length is equal to the max length.
Example: `_.truncate(label, {length: maxSize})`

File: web/ui/src/app/lineage/graph/graph.model.ts [hunk 9]
```diff
@@ -47,16 +47,24 @@ export abstract class VisNode implements vis.Node {
                 public icon: any,
                 public type: VisNodeType) {
     }
+
```
Comment (wajda) [inline web/ui/src/app/lineage/graph/graph.model.ts]: Another redundant method. Please take a look at LoDash `clone()` function. There are four variants of it.

File: web/ui/src/app/lineage/graph/graph.model.ts [hunk 10]
```diff
@@ -70,11 +78,12 @@ export class RegularVisNode extends VisNode {
 }
 
 export class HighlightedVisNode extends VisNode {
-    constructor(public operation: IOperation) {
+    constructor(
+        public operation: IOperation, public label : string) {
```
Comment (wajda) [inline web/ui/src/app/lineage/graph/graph.model.ts]: Minor thing: Line breaks. For a multi-line method/constructor signature, it's a good practice to put every argument on its own line. So you either place everything on one line or one arg per line.

Comment (leborgneromain) [reply web/ui/src/app/lineage/graph/graph.model.ts]: corrected

File: web/ui/src/app/lineage/graph/graph.model.ts (hunk 9 above)
Comment (leborgneromain) [reply web/ui/src/app/lineage/graph/graph.model.ts]: corrected

File: web/ui/src/app/lineage/graph/graph-builder.ts (hunk 8 above)
Comment (leborgneromain) [reply web/ui/src/app/lineage/graph/graph-builder.ts]: corrected

**Comment by wajda** [comment]
Probably the last thing - since you added `package-lock.json` into the repo it should also be removed from the `.gitignore` file.

**Comment by leborgneromain** [comment]
> Probably the last thing - since you added `package-lock.json` into the repo it should also be removed from the `.gitignore` file.

No no that should not be commited. The package-lock.json has nothing to do at the root of the project. this was a wrong npm command launched and stopped manually on this directory. Sorry.

**Comment by wajda** [comment]
LGTM.

> > Probably the last thing - since you added `package-lock.json` into the repo it should also be removed from the `.gitignore` file.
> 
> No no that should not be commited. The package-lock.json has nothing to do at the root of the project. this was a wrong npm command launched and stopped manually on this directory. Sorry.

Actually, and quite surprisingly, according to the NPM doc it is supposed to be committed.
But that's not important for this PR. 

> This file is intended to be committed into source repositories, and serves various purposes:
> https://docs.npmjs.com/files/package-lock.json

**Pull request merged**